components/form.js

```diff
@@ -820,37 +833,39 @@ export function Form ({
     })
   }, [storageKeyPrefix])
 
-  // if `invoiceable` is set,
-  // support for payment per invoice if they are lurking or don't have enough balance
-  // is added to submit handlers.
-  // submit handlers need to accept { satsReceived, hash, hmac } in their first argument
-  // and use them as variables in their GraphQL mutation
-  if (invoiceable && onSubmit) {
-    const options = typeof invoiceable === 'object' ? invoiceable : undefined
-    onSubmit = useInvoiceable(onSubmit, options)
-  }
-
-  const onSubmitInner = useCallback(async (values, ...args) => {
+  const onSubmitInner = useCallback(async ({ amount, ...values }, ...args) => {
+    let cancel, revert, beforeSubmitReturn
     try {
       if (onSubmit) {
-        // extract cost from formik fields
-        // (cost may also be set in a formik field named 'amount')
-        const cost = feeButton?.total || values?.amount
-        if (cost) {
-          values.cost = cost
+        if (requireSession && !me) {
+          throw new SessionRequiredError()
+        }
+        beforeSubmitReturn = beforeSubmit?.({ amount, ...values })
+        let hash, hmac
+        if (invoiceable) {
+          revert = await optimisticUpdate?.({ amount, ...values }, ...args);
+          [{ hash, hmac }, cancel] = await payment.request(amount)
         }
-        await onSubmit(values, ...args)
+        await onSubmit({ hash, hmac, amount, ...values }, ...args)
```

Comment:
I'm only adding hash, hmac, amount to values which was already an object. I don't think this changes the signature.

You might missed the curly braces around the comma separated values?

Reply:
Oh right, values was already an object. What's in args?

Reply:
formik helper stuff like `resetForm`. They call it "formikBag"  in the [documentation](https://formik.org/docs/api/formik#onsubmit-values-values-formikbag-formikbag--void--promiseany). Maybe we should use that name, too. Since we no longer wrap submit handlers for our payment flow, I think `args` is really the same as `formikBag`.

Reply:
Oh I see. I remember using `resetForm` at one point.

Someone at Square told me they don't allow engineers to declare array indices as `i`. Kind of intense lol